### ISSUE TITLE
Invalid code generation for `decode` with `useOptionals=all` + `initializeFieldsAsUndefined=false`

### ISSUE THREAD

**Issue by fxb** [stephenh/ts-proto#736]
When using the options `useOptionals=all` and `initializeFieldsAsUndefined=false` in combination, `ts-proto` generates the following code:

```test.proto
message Test {
  repeated string values = 1;
}
```

```test.ts
function createBaseTest(): Test {
  return {};
}

export const Test = {
  decode(input: _m0.Reader | Uint8Array, length?: number): Test {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseTest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.values!.push(reader.string());  // TypeError: Cannot read properties of undefined (reading 'push')
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  }
};
```

This fails at runtime with `TypeError: Cannot read properties of undefined (reading 'push')`.

To fix this, the affected code snippet should probably not use a non-null assertion operator and instead look as follows:

```test.ts
          message.values = message.values ?? [];
          message.values.push(reader.string());
```

This is also true for maps, which might either use `new Map()` or `{}`.

**Comment by fxb** [comment]
I added a draft PR to fix this issue, which is currently missing test cases. I will convert it to a proper PR once I have added those.

**Comment by fxb** [comment]
Another fix would be to always initialize repeated fields.

**Comment by davidzeng0** [comment]
Any idea when this will be merged?

**Comment by davidzeng0** [comment]
Never mind, just realized you are both the issue author and commit author

**Comment by stephenh** [comment]
@fxb hm, always initializing repeated fields does seem like an easier fix... There is a little risk that existing users are relying on the `initializeFieldsAsUndefined=false` behavior _not_ doing that, but, I dunno...I don't remember if that's on purpose or not.

I'd be cool with trying it and seeing what happens.

@davidzeng0 yeah, we'll see if I can sneak this in, but usually I only have time to give pointers & steward PRs along, instead of diving deep into every issue. The best chance of getting this fixed is if either you or @fxb keep working on it.

Thanks!

**Comment by davidzeng0** [comment]
I think it would be better to initialize upon use as there shouldn't be much of a slowdown + it would be compliant with expected behavior

**Comment by stephenh** [comment]
Fixed in #802 , thanks @davidzeng0 !
